File: backend/Nexus-IMS/src/main/java/com/learn/nexus_ims/services/InvoiceService.java
package com.learn.nexus_ims.services;

import com.learn.nexus_ims.dtos.InvoiceDto;
import com.learn.nexus_ims.entities.Invoice;
import com.learn.nexus_ims.repositories.InvoiceRepository;
import lombok.RequiredArgsConstructor;
import org.modelmapper.ModelMapper;
import org.springframework.stereotype.Service;
import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class InvoiceService {

    private final InvoiceRepository invoiceRepository;
    private final ModelMapper modelMapper;

    /**
     * Create a new invoice
     */
    public InvoiceDto createInvoice(InvoiceDto invoiceDto) {
        Invoice invoice = modelMapper.map(invoiceDto, Invoice.class);
        Invoice savedInvoice = invoiceRepository.save(invoice);
        return modelMapper.map(savedInvoice, InvoiceDto.class);
    }

    /**
     * Get all invoices
     */
    public List<InvoiceDto> getAllInvoices() {
        return invoiceRepository.findAll()
                .stream()
                .map(invoice -> modelMapper.map(invoice, InvoiceDto.class))
                .collect(Collectors.toList());
    }

    /**
     * Get invoice by ID
     */
    public InvoiceDto getInvoiceById(Long id) {
        Invoice invoice = invoiceRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Invoice not found with ID: " + id));
        return modelMapper.map(invoice, InvoiceDto.class);
    }

    /**
     * Update an invoice
     */
    public InvoiceDto updateInvoice(Long id, InvoiceDto invoiceDto) {
        Invoice invoice = invoiceRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Invoice not found with ID: " + id));

        invoice.setInvoiceNo(invoiceDto.getInvoiceNo());
        invoice.setEstimatedId(invoiceDto.getEstimatedId());
        invoice.setChainId(invoiceDto.getChainId());
        invoice.setServiceDetails(invoiceDto.getServiceDetails());
        invoice.setQty(invoiceDto.getQty());
        invoice.setCostPerQty(invoiceDto.getCostPerQty());
        invoice.setAmountPayable(invoiceDto.getAmountPayable());
        invoice.setBalance(invoiceDto.getBalance());
        invoice.setDateOfPayment(invoiceDto.getDateOfPayment());
        invoice.setDateOfService(invoiceDto.getDateOfService());
        invoice.setDeliveryDetails(invoiceDto.getDeliveryDetails());
        invoice.setEmailId(invoiceDto.getEmailId());

        Invoice updatedInvoice = invoiceRepository.save(invoice);
        return modelMapper.map(updatedInvoice, InvoiceDto.class);
    }

    /**
     * Delete an invoice
     */
    public void deleteInvoice(Long id) {
        if (!invoiceRepository.existsById(id)) {
            throw new RuntimeException("Invoice not found with ID: " + id);
        }
        invoiceRepository.deleteById(id);
    }

    /**
     * Get invoices by invoice number
     */
    public InvoiceDto getInvoiceByInvoiceNo(Integer invoiceNo) {
        Invoice invoice = invoiceRepository.findByInvoiceNo(invoiceNo)
                .orElseThrow(() -> new RuntimeException("Invoice not found with Invoice No: " + invoiceNo));
        return modelMapper.map(invoice, InvoiceDto.class);
    }

    /**
     * Get invoices by estimated ID
     */
    public List<InvoiceDto> getInvoicesByEstimatedId(Long estimatedId) {
        return invoiceRepository.findByEstimatedId(estimatedId)
                .stream()
                .map(invoice -> modelMapper.map(invoice, InvoiceDto.class))
                .collect(Collectors.toList());
    }

    /**
     * Get invoices by chain ID
     */
    public List<InvoiceDto> getInvoicesByChainId(Long chainId) {
        return invoiceRepository.findByChainId(chainId)
                .stream()
                .map(invoice -> modelMapper.map(invoice, InvoiceDto.class))
                .collect(Collectors.toList());
    }

    /**
     * Get invoices by email ID
     */
    public List<InvoiceDto> getInvoicesByEmailId(String emailId) {
        return invoiceRepository.findByEmailId(emailId)
                .stream()
                .map(invoice -> modelMapper.map(invoice, InvoiceDto.class))
                .collect(Collectors.toList());
    }

    /**
     * Get invoices by service details
     */
    public List<InvoiceDto> getInvoicesByServiceDetails(String serviceDetails) {
        return invoiceRepository.findByServiceDetails(serviceDetails)
                .stream()
                .map(invoice -> modelMapper.map(invoice, InvoiceDto.class))
                .collect(Collectors.toList());
    }
}

File: frontend/src/services/invoiceService.js
import api from './api';

export const invoiceService = {
  getAllInvoices: async () => {
    const response = await api.get('/invoices');
    return response.data;
  },

  getInvoiceById: async (id) => {
    const response = await api.get(`/invoices/${id}`);
    return response.data;
  },

  createInvoice: async (invoiceData) => {
    const response = await api.post('/invoices', invoiceData);
    return response.data;
  },

  updateInvoice: async (id, invoiceData) => {
    const response = await api.put(`/invoices/${id}`, invoiceData);
    return response.data;
  },

  deleteInvoice: async (id) => {
    await api.delete(`/invoices/${id}`);
  },

  getInvoiceByInvoiceNo: async (invoiceNo) => {
    const response = await api.get(`/invoices/invoiceNo/${invoiceNo}`);
    return response.data;
  },

  getInvoicesByEstimatedId: async (estimatedId) => {
    const response = await api.get(`/invoices/estimated/${estimatedId}`);
    return response.data;
  },

  getInvoicesByChainId: async (chainId) => {
    const response = await api.get(`/invoices/chain/${chainId}`);
    return response.data;
  },

  getInvoicesByEmailId: async (emailId) => {
    const response = await api.get(`/invoices/email/${emailId}`);
    return response.data;
  },

  getInvoicesByServiceDetails: async (serviceDetails) => {
    const response = await api.get(`/invoices/service/${serviceDetails}`);
    return response.data;
  },
};
